tools/codegen/api/lazy.py

```diff
@@ -0,0 +1,175 @@
+        elif str(typ.elem) == 'int':
+            return VectorCType(BaseCType(intT))
+        else:
+            raise AssertionError(f"TODO add support for type {repr(typ)}")
+    else:
+        raise AssertionError(f"unrecognized type {repr(typ)}")
+
+
+def isValueType(typ: Union[Type, BaseCType, OptionalCType, ConstRefCType, MutRefCType, ListCType, ArrayRefCType, ArrayCType, VectorCType, TupleCType]) -> bool:
+    """
+    Given a type, determine if it is a Value-like type.  This is equivalent to
+    being Tensor-like, but assumes the type has already been transformed.
+    """
+    if isinstance(typ, BaseCType):
+        return typ.type == valueT
+    elif isinstance(typ, (OptionalCType, ListCType, VectorCType)):
+        return isValueType(typ.elem)
+    else:
+        return False
+
+# Inspired by a FunctionSchema object, a LazyIrSchema holds the schema of a Lazy IR node.
+# Unlike a FunctionSchema, it has no round-trippable string form (relating to the YAML),
+# but carries type information from a native FunctionSchema modified for use with IR nodes,
+# and preserving original argument names.
+
+
+class LazyIrSchema:
+    # The name of the operator this function schema describes.
+    name: 'OperatorName'
+
+    positional_arg_types: Tuple[NamedCType, ...]
+    keyword_arg_types: Tuple[NamedCType, ...]
+
+    # TODO: Need to handle collisions with argument names at some point
+    returns: Tuple['Return', ...]
+
+    def __init__(self, func: FunctionSchema):
+
+        positional_arg_types = []
+        for arg_field in ["pre_self_positional",
+                          "self_arg",
+                          "post_self_positional"]:
+            if arg_field == "self_arg" and getattr(func.arguments, "self_arg") is not None:
+                arg = getattr(func.arguments, "self_arg").argument
+                positional_arg_types.append(NamedCType(arg.name, process_ir_type(arg.type)))
+            elif getattr(func.arguments, arg_field) is not None:
+                positional_arg_types.extend([
+                    NamedCType(
+                        arg.name,
+                        process_ir_type(arg.type)) for arg in getattr(func.arguments, arg_field)])
+        self.positional_arg_types = tuple(positional_arg_types)
+
+        keyword_arg_types = []
+        for arg_field in ["pre_tensor_options_kwarg_only",
+                          "tensor_options",
+                          "post_tensor_options_kwarg_only",
+                          "out"]:
+            if getattr(func.arguments, arg_field) is not None:
+                keyword_arg_types.extend([
+                    NamedCType(
+                        arg.name,
+                        process_ir_type(arg.type)) for arg in getattr(func.arguments, arg_field)])
+        self.keyword_arg_types = tuple(keyword_arg_types)
+        self.name = func.name
+        self.returns = func.returns
+
+    @property
+    def node_name(self) -> str:
```

Comment:
we could probably add this property to an yaml for cases where we would like to have custom mappings such as `aten::expand`

Reply:
oh, this isn't really for custom names.  its just for capitalized names for following c++ naming 
e.g. aten::mean op gets codegenned IR class named 
class Mean 

and underscores should get split out and each word-part Capitalized...

so i just put it here as a helper for parts of the codegen that use it